File: Agent Marketplace Pattern/autogen_typescript_example/agent_marketplace.ts
// Agent Marketplace Pattern - Autogen TypeScript Example
// To run: npm install && npm run agent-marketplace

import axios from 'axios';
import * as readline from 'readline';
import * as dotenv from 'dotenv';
dotenv.config();

const MISTRAL_API_URL = 'https://api.mistral.ai/v1/chat/completions';
const MISTRAL_API_KEY = process.env.MISTRAL_API_KEY;

interface AgentProfile {
  name: string;
  skill: string;
  cost: number;
}

const agents: AgentProfile[] = [
  { name: 'Alice', skill: 'translation', cost: 5 },
  { name: 'Bob', skill: 'summarization', cost: 3 },
  { name: 'Carol', skill: 'data analysis', cost: 4 }
];

async function marketplace(task: string, requiredSkill: string) {
  // Agents bid for the task if they have the required skill
  const bidders = agents.filter(a => a.skill === requiredSkill);
  if (bidders.length === 0) {
    console.log('No agent available for this skill.');
    return;
  }
  // Pick the lowest cost agent
  const winner = bidders.reduce((prev, curr) => (curr.cost < prev.cost ? curr : prev));
  console.log(`Marketplace: Assigning task to ${winner.name} (Skill: ${winner.skill}, Cost: ${winner.cost})`);
  const prompt = `You are ${winner.name}, an expert in ${winner.skill}. Complete this task: ${task}`;
  const response = await axios.post(
    MISTRAL_API_URL,
    {
      model: 'mistral-tiny',
      messages: [{ role: 'user', content: prompt }],
    },
    {
      headers: {
        'Authorization': `Bearer ${MISTRAL_API_KEY}`,
        'Content-Type': 'application/json',
      },
    }
  );
  console.log(`${winner.name}:`, response.data.choices[0].message.content);
}

const rl = readline.createInterface({
  input: process.stdin,
  output: process.stdout
});

rl.question('Task: ', (taskInput: string) => {
  rl.question('Required skill (translation/summarization/data analysis): ', async (skillInput: string) => {
    try {
      await marketplace(taskInput, skillInput.trim());
    } catch (err) {
      console.error('Error:', err);
    }
    rl.close();
  });
});

File: deprecated/agent-marketplace-pattern/langgraph_python_example/agent_marketplace.py
# Agent Marketplace Pattern - LangGraph Python Example

This example demonstrates the Agent Marketplace Pattern using LangGraph and Python. Agents with different skills and costs bid for a task, and the marketplace assigns the task to the best match. The LLM is Mistral.

## Requirements

- Python 3.8+
- `langgraph` library
- `python-dotenv` (for .env support)
- Mistral LLM API access

## Install dependencies

```bash
pip install langgraph python-dotenv requests
```

## Example Code

```python
import os
from langgraph import Agent, Environment, LLM
from dotenv import load_dotenv

load_dotenv()

MISTRAL_API_KEY = os.getenv("MISTRAL_API_KEY")
MISTRAL_API_URL = "https://api.mistral.ai/v1/chat/completions"

class SimpleEnvironment(Environment):
    def get_observation(self):
        task = input("Task: ")
        skill = input("Required skill (translation/summarization/data analysis): ")
        return task, skill
    def send_action(self, action):
        print(action)

class MarketplaceAgent(Agent):
    def __init__(self, llm, name, skill, cost):
        self.llm = llm
        self.name = name
        self.skill = skill
        self.cost = cost
    def act(self, task):
        prompt = f"You are {self.name}, an expert in {self.skill}. Complete this task: {task}"
        return self.llm.complete(prompt)

llm = LLM(
    provider="mistral",
    api_key=MISTRAL_API_KEY,
    api_url=MISTRAL_API_URL,
)

agents = [
    MarketplaceAgent(llm, "Alice", "translation", 5),
    MarketplaceAgent(llm, "Bob", "summarization", 3),
    MarketplaceAgent(llm, "Carol", "data analysis", 4)
]

env = SimpleEnvironment()
task, required_skill = env.get_observation()
bidders = [a for a in agents if a.skill == required_skill]
if not bidders:
    env.send_action("No agent available for this skill.")
else:
    winner = min(bidders, key=lambda a: a.cost)
    env.send_action(f"Marketplace: Assigning task to {winner.name} (Skill: {winner.skill}, Cost: {winner.cost})")
    result = winner.act(task)
    env.send_action(f"{winner.name}: {result}")
```

---

- Try a task and required skill to see agent bidding and assignment.
- Make sure your `.env` file contains your Mistral API key.
